### PR TITLE
Feature/create chart generator agent

### PR DESCRIPTION
**Pull request by hsauve-scottlogic**
## Description

Implement ChartGeneratorAgent that generates and executes chart code using Matplotlib

## Changelog

- Edit intent agent prompt
- Create ChartGeneratorAgent that generates code, sanitises it and executes it
- Add method to Connection Manager to send chart via websocket
- Edit Director which sends the chart if it exists in the scratchpad
- Add MessageType in frontend
- Edit chat component in the frontend to display message with IMAGE MessageType

Eg. What were mortgage interest rates in 2022 and 2023? Show me the results in a chart

output

Example of questions for the agent:

- What are the different subscriptions with Netflix? Show me the results in a chart
- What was my spending at Tesco, compared to Amazon? Create a chart to show me the results
- What were mortgage interest rates in 2022 and 2023? Show me the results in a chart

### CODE REVIEW

File: backend/src/agents/chart_generator_agent.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            os.makedirs(output_dir)
+        logger.info("Saving the figure as 'output.png'")
+        output_path = os.path.join(output_dir, "output.png")
+        fig.savefig(output_path)
+        logger.info(f"Figure saved successfully as {output_path}")
+
+        with open(output_path, "rb") as image_file:
+            image_data = base64.b64encode(image_file.read()).decode("utf-8")
+
+        await connection_manager.send_chart({"type": "image", "data": image_data})
```
Comment (philbaxter) [inline backend/src/agents/chart_generator_agent.py:47]: Should the agent be sending the chart to the frontend, or, should that be the job of the answer agent, meaning we should probably put the chart on the scratchpad.

File: backend/src/agents/chart_generator_agent.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        exec(sanitised_script, {}, local_vars)
+        fig = local_vars.get('fig')
+        if fig is None:
+            raise ValueError("The generated code did not produce a figure named 'fig'.")
+        output_dir = '/app/output'
+        if not os.path.exists(output_dir):
+            os.makedirs(output_dir)
+        logger.info("Saving the figure as 'output.png'")
+        output_path = os.path.join(output_dir, "output.png")
+        fig.savefig(output_path)
```
Comment (philbaxter) [inline backend/src/agents/chart_generator_agent.py:41]: Why dont we just keep this in memory? I also dont see anywhere where we clean up the file we have created once we are done with it.

**Review by philbaxter — CHANGES_REQUESTED** [review]
Good work, a couple of comments.

File: backend/requirements.txt [hunk 3]
```diff
@@ -17,7 +17,10 @@ cryptography==42.0.7
 isodate==0.6.1
 pycparser==2.22
 openai==1.35.3
+matplotlib==3.9.1
 beautifulsoup4==4.12.3
 aiohttp==3.9.5
 googlesearch-python==1.2.4
+matplotlib==3.9.1
 pyyaml==6.0.2
+pillow
```
Comment (gaganahluwalia) [inline backend/requirements.txt:26]: Please include the version of this library.

File: backend/src/agents/chart_generator_agent.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        exec(sanitised_script, {}, local_vars)
+        fig = local_vars.get('fig')
+        buf = BytesIO()
+        if fig is None:
+            raise ValueError("The generated code did not produce a figure named 'fig'.")
+        fig.savefig(buf, format='png')
+        buf.seek(0)
+        with Image.open(buf):
+            image_data = base64.b64encode(buf.getvalue()).decode("utf-8")
+
```
Comment (gaganahluwalia) [inline backend/src/agents/chart_generator_agent.py:40]: Pleas delete this extra line.

File: backend/src/agents/chart_generator_agent.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+def sanitise_script(script: str) -> str:
+    script = script.strip()
+    if script.startswith("```python"):
+        script = script[9:]
+    if script.endswith("```"):
+        script = script[:-3]
+    return script.strip()
+
+@tool(
```
Comment (gaganahluwalia) [inline backend/src/agents/chart_generator_agent.py:56]: Please include model and llm as parameters as well.

**Review by gaganahluwalia — CHANGES_REQUESTED** [review]
As your PR includes the changes related to env.yml to .env file. Could you please revert the changes made in the main README.md file as well.

**Review by gaganahluwalia — APPROVED** [review]

**Review by gaganahluwalia — APPROVED** [review]

**Review by SvetA1395 — APPROVED** [review]

**Pull request merged**